src/main/java/seedu/address/history/HistoryManager.java

```diff
@@ -89,12 +89,14 @@ public void addToHistory(String command) {
         hasReturnedCurrentCommandBefore = false;
         if (hasNextCommand()) {
             // if there exist a next command,
-            // copy commandHistory[0] to commandHistory[currentCommandIndex],
-            // store it as the new commandHistory
-            // and append the new command.
-            commandHistory = commandHistory.subList(0, currentCommandIndex + 1);
-            this.currentCommandIndex++;
+            // append the new command and
+            // shift the currentCommandHistoryIndex to the end of the list
             this.commandHistory.add(command);
+            if (isLimitReached()) {
+                // if adding the new command hits the limit, pop the first command off
+                this.commandHistory.remove(0);
+            }
+            this.currentCommandIndex = commandHistory.size() - 1;
         } else if (isLimitReached()) {
             this.commandHistory.remove(0);
             this.commandHistory.add(command);
```

Comment:
```suggestion
            // shift the currentCommandHistoryIndex to the end of the list
            this.currentCommandIndex = commandHistory.size() - 1;
        }
         if (isLimitReached()) {
            this.commandHistory.remove(0);
```

Then below this in the else

```else {
    this.currentCommandIndex++;
}

this.commandHistory.add(command)
```

This makes the happy path much much clearer